### ISSUE TITLE
make: *** No rule to make target 'all' although  "CMake Make Builder" is selected

### ISSUE THREAD

**Issue by akors** [15knots/cmake4eclipse#50]
Hi, I've been trying to use cmake4eclipse in Eclipse Mars for a project.

In Project -> Properties -> C/C++ Build -> Tool Chain Editor, I definitely selected "CMake Make Builder",
And In C/C++ Build, I have selected the External Builder, and "Generate Makefiles automatically".

However, Eclipse still pretends that I'm building a Make Project:
```
12:30:56 **** Incremental Build of configuration Debug for project xpscore ****
make all 
make: *** No rule to make target 'all'.  Stop.
```

When I click on Build Project, the Progress window also shows "Invoking CDT Builder", nothing about CMake. I don't know what I'm doing wrong though. How can I get Eclipse to use CMake?

**Comment by 15knots** [comment]
Looks like you did not active the plugin for building at all. The output under linux should read:
```
13:00:08 **** Incremental Build of configuration Debug for project C-1src ****
/usr/bin/gmake all 
Scanning dependencies of target EXECUTABLE
[ 50%] Building C object CMakeFiles/EXECUTABLE.dir/main.c.o
[100%] Linking C executable EXECUTABLE
[100%] Built target EXECUTABLE
```

In the console view, do you see an entry `CMake Console` in the console display menu? If not, CDT is definitely using its own builder.
You might also prefer the _CMake Builder (portable)_, since it is not hard-coded to use ```make``` but uses the build tool that cmake detected, which allows you to build with ```ninja```

**Comment by akors** [comment]
> If not, CDT is definitely using its own builder.

Yes it is, but WHY is it using its own builder? Here is my Tool Chain Editor tab:

![bildschirmfoto von 2017-10-23 15-07-06](https://user-images.githubusercontent.com/3023492/31890606-00236cb6-b804-11e7-94e4-1ae652cdfb18.png)

**Comment by 15knots** [comment]
> Yes it is, but WHY is it using its own builder?

When you select a different CDT tool-chain, CDT internally switches the builder to the builder defined in the CDT tool-chain, which disables the Cmake Make builder. You will notice that the field 'C/C++Build | Builder Settings | Build directory' shows sth. different than `/xpscore/build/Debug` in your case.
> How can I get eclipse to use the cmake builder?

1. Set the current toolchain to `Linux GCC` or whatever is the selected as the default. The information from the CDT toolchain is only [1] relevant, if CDT generates the build-scripts. With cmake4eclipse
, cmake will generate them, it has its own mechanism to specify its tool-chain.
2. Select `CMake Builder (portable)` as the current builder.
3. Tell cmake to use CUDA Linux toolchain by passing a toolchain file argument to cmake: Add a `CMAKE_TOOLCHAIN_FILE` entry on the Symbols tab; see  [details](https://cmake.org/cmake/help/v3.6/manual/cmake-toolchains.7.html#cross-compiling) on cmake.org.

[1] Pro tip: Leave at least an entry for the C-compiler and the C++-compiler in the CDT toolchain tools list, otherwise the CDT indexer will not work properly.

**Comment by akors** [comment]
Getting another error now:

```
Errors running builder 'CDT Builder' on project 'myproject'.
A resource already exists on disk '/path/to/myproject/build'.
A resource already exists on disk '/path/to/myproject/build'.
```

Yes, current Toolchain is set to `Linux GCC` and current builder is `CMake Builder (portable)`.
The directory already exists, but that shouldn't matter, should it?

**Comment by 15knots** [comment]
> The directory already exists, but that shouldn't matter, should it?

No, it should not matter.
Where do you see the message `Errors running builder 'CDT Builder' on project 'myproject'.`? In the build log?

**Comment by akors** [comment]
@15knots 

It worked for a while for me, but after a few version upgrades (to Eclipse 4.7.2 and cmake4eclipse 1.10.0),  I keep having the same problem: 

```
12:30:56 **** Incremental Build of configuration Debug for project xpscore ****
make all 
make: *** No rule to make target 'all'.  Stop.
```

I have selected another builder and then the `CMake Builder (portable)` again, but eclipse just refuses to use it. I don't know of another method of selecting the CMake builder other than selecting it in the project menu.

How do I force Eclipse to use the `CMake Builder (portable)` builder?

**Comment by akors** [comment]
When I went into the C++-Build Menu and **de**selected "Generate Makefiles automatically" under the "Makefile Generation" section of the "Builder Settings" and manually selected the build directory, it started working.

Is this normal behaviour? Seems a bit non-obvious to me.

**Comment by 15knots** [comment]
> 12:30:56 **** Incremental Build of configuration Debug for project xpscore ****

I always get '**Auto** Build of ...', looks like the CDT builder kicks in again.

> make all
If you see a line like `CMAKE_MAKE_PROGRAM:FILEPATH=/usr/bin/gmake` (note the full path to gmake) in file `CMakeCache.txt` in the build directory **and** the commandline is just
`make all`, it is definitely the CDT builder.

**Comment by 15knots** [comment]
Concerning https://github.com/15knots/cmake4eclipse/issues/50#issuecomment-366947206: If you disable makefile generation, the plugin will no longer invoke cmake.
The build itself may or may not work, that is out of the plugins scope.

**Comment by akors** [comment]
> Concerning #50 (comment): If you disable makefile generation, the plugin will no longer invoke cmake.
The build itself may or may not work, that is out of the plugins scope.

OK, I understand. The build only worked because I already had a previously configured project there, not because the Project is actually correctly generated, correct?

> If you see a line like CMAKE_MAKE_PROGRAM:FILEPATH=/usr/bin/gmake (note the full path to gmake) in file CMakeCache.txt in the build directory and the commandline is just
make all, it is definitely the CDT builder.
The commandline should be /usr/bin/gmake all

I see `CMAKE_MAKE_PROGRAM:FILEPATH=/usr/bin/gmake` in the CMakeCache.txt, but the command line is `/usr/bin/gmake -j8 all`. Considering I set the bulid command to `CMAKE_BUILD_TOOL -j8` in Project -> Properties -> C/C++ Build, I'm going to assume that it's using the right builder.


So I removed my build directory, and all of a sudden the project builds again. Man, this bug is confusing. 
 :confounded:

Sorry for the back-and-forth, but that's what I'm experiencing here :/
